net: DHCPv4: Filter NAK during requesting state

rerickson1 opened the pull request:
During the requesting state, ignore NAKs from any
non-requesting server.  This gives the requesting
server a chance to ACK the request.

fixes #50868

rerickson1 commented:
See logs below proving the fix.
DHCP server 1 = 192.168.210.1
DHCP server 2 = 192.168.84.1
```
[00:00:01.717,712] <inf> ethdrv: Link up
[00:00:01.717,956] <inf> ethdrv: 100Mbps
[00:00:01.717,956] <inf> ethdrv: full duplex
[00:00:01.718,139] <inf> ethdrv: Not suspended
[00:00:01.718,200] <dbg> net_dhcpv4: dhcpv4_iface_event_handler: (net_mgmt): Interface 0x20004018 coming up
[00:00:01.718,231] <dbg> net_dhcpv4: dhcpv4_immediate_timeout: (net_mgmt): force timeout dhcpv4=0x20004028
[00:00:01.718,536] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=init timeleft=0
[00:00:01.718,566] <dbg> net_dhcpv4: dhcpv4_enter_selecting: (sysworkq): enter state=selecting
[00:00:01.720,794] <dbg> net_dhcpv4: dhcpv4_set_timeout: (sysworkq): sched timeout dhcvp4=0x20004028 timeout=5s
[00:00:01.720,825] <dbg> net_dhcpv4: dhcpv4_send_discover: (sysworkq): send discover xid=0xdaaee20c timeout=5s
[00:00:01.720,855] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 5s
[00:00:01.721,069] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=selecting timeleft=5
[00:00:01.721,099] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 5s
[00:00:06.091,125] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0x2d2a41e0 secs=0 flags=0x80 chaddr=A2:5F:A0:4B:A4:69
[00:00:06.091,156] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:06.091,156] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=0.0.0.0
[00:00:06.091,186] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=0.0.0.0
[00:00:06.091,217] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:06.091,247] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Unexpected op (2), xid (daaee20c vs 2d2a41e0) or chaddr
[00:00:06.721,191] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=selecting timeleft=0
[00:00:06.723,480] <dbg> net_dhcpv4: dhcpv4_set_timeout: (sysworkq): sched timeout dhcvp4=0x20004028 timeout=7s
[00:00:06.723,510] <dbg> net_dhcpv4: dhcpv4_send_discover: (sysworkq): send discover xid=0xdaaee20d timeout=7s
[00:00:06.723,541] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 7s
[00:00:06.723,785] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=selecting timeleft=7
[00:00:06.723,815] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 7s
[00:00:06.724,456] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0xdaaee20d secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:06.724,487] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:06.724,487] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=192.168.210.193
[00:00:06.724,517] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=192.168.210.1
[00:00:06.724,548] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:06.724,670] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.210.1
[00:00:06.724,700] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_lease_time: 86400
[00:00:06.724,731] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_renewal: 43200
[00:00:06.724,761] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_rebinding: 75600
[00:00:06.724,822] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_subnet_mask 255.255.255.0
[00:00:06.724,853] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 28
[00:00:06.725,006] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_router: 192.168.210.1
[00:00:06.725,036] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:06.725,128] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=selecting msg=offer
[00:00:06.725,158] <dbg> net_dhcpv4: dhcpv4_enter_requesting: (rx_q[0]): enter state=requesting
[00:00:06.725,189] <dbg> net_dhcpv4: dhcpv4_set_timeout: (rx_q[0]): sched timeout dhcvp4=0x20004028 timeout=3s
[00:00:06.727,569] <dbg> net_dhcpv4: dhcpv4_send_request: (rx_q[0]): send request dst=255.255.255.255 xid=0xdaaee20e ciaddr=<unknown> +server-id +requested-ip timeout=3s
[00:00:06.727,691] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=requesting timeleft=3
[00:00:06.727,691] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 3s
[00:00:06.728,729] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0xdaaee20e secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:06.728,759] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:06.728,790] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=0.0.0.0
[00:00:06.728,790] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=0.0.0.0
[00:00:06.728,820] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:06.728,912] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.84.1
[00:00:06.728,942] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 56
[00:00:06.728,973] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:06.729,095] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=requesting msg=nak
[00:00:06.734,619] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0xdaaee20e secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:06.734,649] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:06.734,680] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=192.168.210.193
[00:00:06.734,710] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=192.168.210.1
[00:00:06.734,741] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:06.734,832] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.210.1
[00:00:06.734,863] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_lease_time: 86400
[00:00:06.734,924] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_renewal: 43200
[00:00:06.734,954] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_rebinding: 75600
[00:00:06.735,015] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_subnet_mask 255.255.255.0
[00:00:06.735,046] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 28
[00:00:06.735,076] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 119
[00:00:06.735,229] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_router: 192.168.210.1
[00:00:06.735,260] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:06.735,351] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=requesting msg=ack
[00:00:06.735,382] <inf> net_dhcpv4: Received: 192.168.210.193
[00:00:06.735,443] <dbg> net_dhcpv4: dhcpv4_enter_bound: (rx_q[0]): enter state=bound renewal=43200s rebinding=75600s
[00:00:06.735,443] <dbg> net_dhcpv4: dhcpv4_set_timeout: (rx_q[0]): sched timeout dhcvp4=0x20004028 timeout=43200s
[00:00:06.735,565] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=bound timeleft=43200
[00:00:06.735,595] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 43200s
[00:00:07.233,276] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0x2d2a41e1 secs=256 flags=0x80 chaddr=A2:5F:A0:4B:A4:69
[00:00:07.233,306] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:07.233,337] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=0.0.0.0
[00:00:07.233,337] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=0.0.0.0
[00:00:07.233,367] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:07.233,398] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Unexpected op (2), xid (daaee20e vs 2d2a41e1) or chaddr

```

nordicjm commented:
This does not look valid according to RFC2131:

      If the client receives a DHCPNAK message, it cannot reuse its
      remembered network address.  It must instead request a new
      address by restarting the configuration process, this time
      using the (non-abbreviated) procedure described in section
      3.1.  This action also corresponds to the client moving to
      the INIT state in the DHCP state diagram.

rerickson1 commented:
Yeah, I think this solution I proposed is not exactly correct.
There is an issue with the Zephyr implementation though. I have over 70 devices on my network and the Zephyr one is the only one that has an issue an spins forever not getting an address.
I think the proper solution here is the client should keep track of the server address it is requesting for and ignore naks for other servers in the requesting state.
See Section 4.4 of the RFC. It specifically has a point where the client can choose which server to request from. Once it requests, it should only care about responses from that server, not others. If at that point it gets a nak from the server it requests to, then it should start over.

rlubos commented:
> I think the proper solution here is the client should keep track of the server address it is requesting for and ignore naks for other servers in the requesting state.
See Section 4.4 of the RFC. It specifically has a point where the client can choose which server to request from. Once it requests, it should only care about responses from that server, not others. If at that point it gets a nak from the server it requests to, then it should start over.

Sounds reasonable to me.

rerickson1 commented:
Updated to filter NAKs appropriately.

Log output showing it work properly:

```
[00:00:05.593,872] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=selecting timeleft=0
[00:00:05.595,947] <dbg> net_dhcpv4: dhcpv4_set_timeout: (sysworkq): sched timeout dhcvp4=0x20004028 timeout=8s
[00:00:05.596,008] <dbg> net_dhcpv4: dhcpv4_send_discover: (sysworkq): send discover xid=0x715711f6 timeout=8s
[00:00:05.596,038] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 8s
[00:00:05.596,221] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=selecting timeleft=8
[00:00:05.596,252] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 8s
[00:00:05.597,167] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0x715711f6 secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:05.597,167] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:05.597,198] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=192.168.210.193
[00:00:05.597,229] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=192.168.210.1
[00:00:05.597,229] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:05.597,320] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.210.1
[00:00:05.597,351] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_lease_time: 86400
[00:00:05.597,412] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_renewal: 43200
[00:00:05.597,442] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_rebinding: 75600
[00:00:05.597,473] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_subnet_mask 255.255.255.0
[00:00:05.597,503] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 28
[00:00:05.597,656] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_router: 192.168.210.1
[00:00:05.597,686] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:05.597,778] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=selecting msg=offer
[00:00:05.597,808] <dbg> net_dhcpv4: dhcpv4_enter_requesting: (rx_q[0]): enter state=requesting
[00:00:05.597,839] <dbg> net_dhcpv4: dhcpv4_set_timeout: (rx_q[0]): sched timeout dhcvp4=0x20004028 timeout=4s
[00:00:05.600,158] <dbg> net_dhcpv4: dhcpv4_send_request: (rx_q[0]): send request dst=255.255.255.255 xid=0x715711f7 ciaddr=<unknown> +server-id +requested-ip timeout=4s
[00:00:05.600,219] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=requesting timeleft=4
[00:00:05.600,250] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 4s
[00:00:05.601,287] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0x715711f7 secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:05.601,318] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:05.601,318] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=0.0.0.0
[00:00:05.601,348] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=0.0.0.0
[00:00:05.601,379] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:05.601,470] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.84.1
[00:00:05.601,501] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 56
[00:00:05.601,501] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:05.601,623] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=requesting msg=nak
[00:00:05.601,654] <dbg> net_dhcpv4: dhcpv4_handle_msg_nak: NAK from non-requesting server 192.168.84.1, ignore it
[00:00:05.605,957] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]): Received dhcp msg [op=0x2 htype=0x1 hlen=6 xid=0x715711f7 secs=0 flags=0x80 chaddr=E8:EB:1B:CD:B2:42
[00:00:05.605,987] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   ciaddr=0.0.0.0
[00:00:05.605,987] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   yiaddr=192.168.210.193
[00:00:05.606,018] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   siaddr=192.168.210.1
[00:00:05.606,048] <dbg> net_dhcpv4: net_dhcpv4_input: (rx_q[0]):   giaddr=0.0.0.0]
[00:00:05.606,140] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_server_id: 192.168.210.1
[00:00:05.606,170] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_lease_time: 86400
[00:00:05.606,231] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_renewal: 43200
[00:00:05.606,262] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_rebinding: 75600
[00:00:05.606,323] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_subnet_mask 255.255.255.0
[00:00:05.606,353] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 28
[00:00:05.606,384] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): option unknown: 119
[00:00:05.606,506] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_router: 192.168.210.1
[00:00:05.606,536] <dbg> net_dhcpv4: dhcpv4_parse_options: (rx_q[0]): options_end
[00:00:05.606,658] <dbg> net_dhcpv4: dhcpv4_handle_reply: (rx_q[0]): state=requesting msg=ack
[00:00:05.606,689] <inf> net_dhcpv4: Received: 192.168.210.193
[00:00:05.606,719] <dbg> net_dhcpv4: dhcpv4_enter_bound: (rx_q[0]): enter state=bound renewal=43200s rebinding=75600s
[00:00:05.606,750] <dbg> net_dhcpv4: dhcpv4_set_timeout: (rx_q[0]): sched timeout dhcvp4=0x20004028 timeout=43200s
[00:00:05.606,872] <dbg> net_dhcpv4: dhcpv4_manage_timers: (sysworkq): iface 0x20004018 dhcpv4=0x20004028 state=bound timeleft=43200
[00:00:05.606,903] <dbg> net_dhcpv4: dhcpv4_timeout: (sysworkq): Waiting for 43200s
```

cfriedt commented:
@rerickson1 - the test suite for dhcpv4 seems like it could use a bit of attention. Just throwing that out there 😬

rerickson1 commented:
> @rerickson1 - the test suite for dhcpv4 seems like it could use a bit of attention. Just throwing that out there 😬

I see that... I think that's a task for another day 😉

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -941,10 +945,20 @@ static void dhcpv4_handle_msg_nak(struct net_if *iface)
 	case NET_DHCPV4_INIT:
 	case NET_DHCPV4_SELECTING:
+	case NET_DHCPV4_REQUESTING:
+		if (memcmp(&request_server_addr, &iface->config.dhcpv4.server_id,
```
I'm not sure if such an comparision is roboust enough. I see in RFC2132, that Server ID option, which updates the `config.dhcpv4.server_id` field is optional for DHCPNAK message. It could be the case, that the "wrong" server sends a NAK w/o this option, yet it'd still be processed. IMO it'd be better to rely on the source address present in the IP header.

And generally, probably not something to be fixed in this PR, but I find it actually awkward to process the DHCP options before it's decided whether to process or drop the packet.

rerickson1 replied:
I think this works sufficiently. The key here is that in the selecting state, an offer message has to have the server id option set. So `server_id` will be the offer server which is what we want. If it is a nak, it will be empty which won't match or it will be the address of the naking server.

rerickson1 replied:
Not sure what I was thinking, I see your point now. I fixed this. The src address of a response is used instead of the server_id field.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -39,6 +39,8 @@ static struct k_work_delayable timeout_work;
 
 static struct net_mgmt_event_callback mgmt4_cb;
 
+static struct in_addr request_server_addr;
```
Shoudn't this be a part of `iface->config.dhcpv4`? A global variable approach won't work well in case there are mutliple interfaces running DHCP. Ideally I think we could rely on `server_id` instead, but that'd require to sort out the option processing first.

rerickson1 replied:
Fixed

rerickson1 committed:
net: DHCPv4: Filter NAK during requesting state

During the requesting state, ignore NAKs from any
non-requesting server.  This gives the requesting
server a chance to ACK the request.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -332,6 +332,12 @@ struct net_if_dhcpv4 {
 	uint8_t attempts;
+
+	/** The address of the server the request is sent to */
+	struct in_addr request_server_addr;
```
Do we need this address, would server_id field store the same data anyway?

rerickson1 replied:
Yes I believe we do. The server_id field can get changed by a received response so then we wouldn't be comparing against the original request address.

jukkar replied:
Ok, I was reading the code but it was not very clear whether it is overwritten or not.
